pkg/vault/contracts/BatchRouterHooks.sol

```diff
@@ -381,163 +432,237 @@ contract BatchRouterHooks is BatchRouterCommon {
+                userData: userData
+            })
+        );
+
+        if (isLastStep) {
+            // Refund unused portion of BPT to the user.
+            if (amountIn < stepMaxAmountIn && sender != address(this)) {
+                stepTokenIn.safeTransfer(address(sender), stepMaxAmountIn - amountIn);
+            }
+        } else if (amountIn < stepMaxAmountIn) {
+            // Refund unused portion of BPT flashloan to the Vault.
+            uint256 refundAmount = stepMaxAmountIn - amountIn;
+            stepTokenIn.safeTransfer(address(_vault), refundAmount);
+            _vault.settle(stepTokenIn, refundAmount);
+        }
+
+        if (isLastStep) {
+            _settledTokenAmounts().tAdd(address(stepTokenIn), amountIn);
+        }
+    }
+
+    function _addLiquidityExactOut(
+        bytes memory userData,
+        address sender,
+        address pool,
+        IERC20 stepTokenIn,
+        IERC20 tokenOut,
+        uint256 stepExactAmountOut,
+        uint256 stepMaxAmountIn,
+        bool isFirstStep,
+        bool isLastStep
+    ) internal returns (uint256 amountIn) {
+        // Token out is BPT: add liquidity - Single token exact out.
+        (uint256[] memory stepAmountsIn, uint256 tokenIndex) = _getSingleInputArrayAndTokenIndex(
+            pool,
+            stepTokenIn,
+            stepMaxAmountIn
+        );
+
+        // Reusing `amountsIn` as input argument and function output to prevent stack too deep error.
+        (stepAmountsIn, , ) = _vault.addLiquidity(
+            AddLiquidityParams({
+                pool: pool,
+                to: isFirstStep ? sender : address(_vault),
+                maxAmountsIn: stepAmountsIn,
+                minBptAmountOut: stepExactAmountOut,
+                kind: AddLiquidityKind.SINGLE_TOKEN_EXACT_OUT,
+                userData: userData
+            })
+        );
+
+        amountIn = stepAmountsIn[tokenIndex];
+
+        uint256 amountOut = isLastStep ? stepExactAmountOut : amountIn;
+
+        // The first step executed determines the outputs for the path, since this is given out.
+        if (isFirstStep) {
+            // Instead of sending tokens back to the Vault, we can just discount it from whatever
+            // the Vault owes the sender to make one less transfer.
+            _currentSwapTokenOutAmounts().tSub(address(tokenOut), amountOut);
+        } else {
+            // If it's not the first step, BPT is minted to the Vault so we just get the credit.
+            _vault.settle(IERC20(pool), amountOut);
+        }
+
+        if (isLastStep) {
+            _currentSwapTokenInAmounts().tAdd(address(stepTokenIn), amountIn);
+        }
+    }
+
+    function _swapExactOut(
+        bytes memory userData,
+        address pool,
+        IERC20 stepTokenIn,
+        IERC20 stepTokenOut,
+        uint256 stepExactAmountOut,
+        uint256 stepMaxAmountIn,
+        bool isLastStep
+    ) internal returns (uint256 amountIn) {
```

Comment:
very silly comment (here and elsewhere): I'm a bit curious about the order of the parameters.

I'd probably have chosen to align the parameters with how they're used inside the function. Then, you'd have `pool`, `stepTokenIn`, `stepTokenOut`, `stepExactAmountOut`, `stepMaxAmountIn` and finally `userData` (usually `userData` goes at the end). Any particular reason to change that?

Reply:
Yes, it’s a special hack to avoid hitting the “stack too deep” issue and to save gas. In the usual ordering, we run into “stack too deep,” and the solution ends up consuming a lot of gas.